crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+			let origin = ensure_signed(origin)?;
+
+			ensure!(!collection_id.is_empty() && collection_id.len() <= MAX_COLLECTION_ID_LENGTH as usize, Error::<T>::CollectionIdInvalid);
+			ensure!(core::str::from_utf8(&collection_id).is_ok(), Error::<T>::CollectionIdInvalid);
+			ensure!(!<CollectionOwner<T>>::contains_key(&collection_id), Error::<T>::CollectionIdExists);
+
+			ensure!(!schema.is_empty(), Error::<T>::SchemaEmpty);
+			ensure!(schema.len() <= MAX_SCHEMA_FIELDS as usize, Error::<T>::SchemaMaxAttributes);
+
+			// Check the provided attribute types are valid
+			ensure!(
+				schema.iter().all(|(_name, type_id)| NFTAttributeValue::is_valid_type_id(*type_id)),
+				Error::<T>::SchemaInvalid
+			);
+
+			// Attribute names must be unique (future proofing for map lookups etc.)
+			let (attribute_names, _): (Vec<NFTAttributeName>, Vec<NFTAttributeTypeId>) = schema.iter().cloned().unzip();
+			let deduped = BTreeSet::from_iter(attribute_names);
+			ensure!(deduped.len() == schema.len(), Error::<T>::SchenmaDuplicateAttribute);
+
+			// Create the collection, update ownership, and bookkeeping
+			if let Some(royalties_schedule) = royalties_schedule {
+				ensure!(royalties_schedule.validate(), Error::<T>::RoyaltiesOvercommitment);
+				<CollectionRoyalties<T>>::insert(&collection_id, royalties_schedule);
+			}
+			CollectionSchema::insert(&collection_id, schema);
+			<CollectionOwner<T>>::insert(&collection_id, origin.clone());
+
+			Self::deposit_event(RawEvent::CreateCollection(collection_id, origin));
+
+			Ok(())
+		}
+
+		/// Issue a new NFT
+		/// `owner` - the token owner
+		/// `attributes` - initial values according to the NFT collection/schema
+		/// `royalties_schedule` - optional royalty schedule for secondary sales of _this_ token, defaults to the collection config
+		/// Caller must be the collection owner
+		#[weight = T::WeightInfo::create_token()]
+		#[transactional]
+		fn create_token(origin, collection_id: CollectionId, owner: T::AccountId, attributes: Vec<NFTAttributeValue>, royalties_schedule: Option<RoyaltiesSchedule<T::AccountId>>) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			// Permission and existence check
+			let collection_owner = Self::collection_owner(&collection_id);
+			ensure!(collection_owner.is_some(), Error::<T>::NoCollection);
+			ensure!(collection_owner.unwrap() == origin, Error::<T>::NoPermission);
+
+			// Check we can issue a new token
+			let token_id = Self::next_token_id(&collection_id);
+			let next_token_id = token_id.checked_add(&One::one()).ok_or(Error::<T>::NoAvailableIds)?;
+			Self::token_issuance(&collection_id).checked_add(&One::one()).ok_or(Error::<T>::MaxTokensIssued)?;
+
+			// Quick `attributes` sanity checks
+			ensure!(!attributes.is_empty(), Error::<T>::SchemaEmpty);
+			ensure!(attributes.len() as u32 <= MAX_SCHEMA_FIELDS, Error::<T>::SchemaMaxAttributes);
```

Comment:
I suggest the above two checks to precede the storage read `Self::collection_owner` at line 270, so we avoid 2 reads when there is an error in the input param.